src/recorder.rs

```diff
@@ -0,0 +1,71 @@
+//! Parser extensions that inspect the input without modifying it.
+//!
+//! *"Only one man stood and watched the sky, stood with terrible sadness in his eyes
+//! and rubber bungs in his ears. He knew exactly what was happening and had known
+//! ever since his Sub-Etha Sens-O-Matic had started winking in the dead of night
+//! beside his pillar and woken him with a start."*
+use crate::{input::Marker, Input};
+use core::ops::{Deref, DerefMut};
+
+#[allow(unused)] // for intra-doc links
+use crate::Parser;
+
+/// A type that receives event hooks when certain parsing actions occur.
```

Comment:
I think I'd be tempted to just call this type `State` and remove the existing type alias(es) (`State`, `Context`, and `Err`). `Recorder` seems too specific a name for what this might eventually turn into.

Reply:
i'm ok with renaming but "State" sounds too generic, i was hoping we could indicate that this is receiving events somehow. i've renamed it to `Inspector` for now.